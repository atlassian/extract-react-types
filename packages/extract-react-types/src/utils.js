/* eslint-disable import/prefer-default-export */

export function hasTypeAnnotation(path, left, right) {
  if (!path.isVariableDeclarator() || !path.node.id.typeAnnotation) {
    return false;
  }

  const { typeName, typeParameters } = path.node.id.typeAnnotation.typeAnnotation;

  if (!typeParameters) return false;

  if (
    (typeName.left && typeName.left.name === left && typeName.right.name === right) ||
    ((right && typeName.name === right) || typeName.name === left)
  ) {
    return true;
  }

  return false;
}
