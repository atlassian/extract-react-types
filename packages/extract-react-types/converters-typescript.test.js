// @flow

/*::
type TestCase = {
  name: string,
  typeSystem: 'flow' | 'typescript',
  code: string,
  only?: boolean,
  skip?: boolean
}
*/

import stripIndent from 'strip-indent';
import cases from 'jest-in-case';
import { extractReactTypes } from './src';

const TESTS = [
  {
    name: 'type comments',
    typeSystem: 'typescript',
    code: `
      interface Props {
        /* Type comment for a */
        a: string;
      }

      class Component extends React.Component<Props> {
      }
    `
  },
  {
    name: 'React.ComponentType',
    typeSystem: 'typescript',
    code: `
    import React from 'react';

    type Props = {
      children: React.ComponentType,
    };

    class Component extends React.Component<Props> {

    }
  `
  },
  {
    name: 'React.FC',
    typeSystem: 'typescript',
    code: `
      type Props = {
        children: string;
      };

      const Component: FC<Props> = (props) => null;

      export default Component;
    `
  },
  {
    name: 'React.FC with empty type argument',
    typeSystem: 'typescript',
    code: `
      type Props = {
        children: string;
      };

      const Component: FC = (props) => null;

      export default Component;
    `
  },
  {
    name: 'React.FC with inline type argument',
    typeSystem: 'typescript',
    code: `
      const Component: FC<{
        children: string;
      }> = (props) => null;

      export default Component;
    `
  },
  {
    name: 'Direct type assignment',
    typeSystem: 'typescript',
    code: `
    interface BadgeProps {
      texture: string;
    }

    function Badge({ texture }: BadgeProps) {}

    Badge.f = [];

    export default Badge;
  `
  },
  {
    name: 'ts nested prop',
    typeSystem: 'typescript',
    code: `
    interface BadgeProps {
      texture: Texture["src"];
    }

    function Badge({ texture }: BadgeProps) {}

    export default Badge;
  `
  },
  {
    name: 'ts array prop',
    typeSystem: 'typescript',
    code: `
    interface ScheduleProps {
      intervals: Array<{
        begin: Interval["begin"];
        end: Interval["end"];
      }>;
    }

    function Schedule({ intervals }: ScheduleProps) {}

    export default Schedule;
  `
  },
  {
    name: 'ts decorators',
    typeSystem: 'typescript',
    code: `
    @ObjectType()
    export class Theme extends React.Component<{}, {}> {
      @Field(_ => ID)
      public id!: string;

      @Field(_ => Textures)
      public fonts!: Textures;

      render() {}
    }

    export default Theme;
  `
  },
  {
    name: 'ts boolean',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{ foo: boolean }> {

    }
  `
  },
  {
    name: 'ts string',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{ foo: string }> {

    }
  `
  },
  {
    name: 'ts as expression',
    typeSystem: 'typescript',
    code: `
    type Foo = 'foo' | 'bar';
    type Props = { bar: Foo }

    class Component extends React.Component<Props> {
      static defaultProps = {
        bar: 'foo' as Foo,
      }
    }
  `
  },
  {
    name: 'ts object',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: { name: string }, bar: number, verified: boolean}> {

    }
  `
  },
  {
    name: 'ts interface',
    typeSystem: 'typescript',
    code: `
    interface ComponentProps {
      foo: string;
      bar: number;
    }

    class Component extends React.Component<ComponentProps> {
      // ...
    }
  `
  },
  {
    name: 'ts function',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: (string, boolean) => number}> {
      // ...
    }
  `
  },
  {
    name: 'ts array',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: Array<number>}> {
      // ...
    }
  `
  },
  {
    name: 'ts union',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: number | string}> {
      // ...
    }
  `
  },
  {
    name: 'ts any',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: any}> {
      // ...
    }
  `
  },
  {
    name: 'ts literals',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: '25' | 30}> {

    }
  `
  },
  {
    name: 'ts optional',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo?: string}> {

    }
  `
  },
  {
    name: 'ts void',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: void}> {

    }
  `
  },
  {
    name: 'ts no react component',
    typeSystem: 'typescript',
    code: `
    class Component extends Foo<{foo: void}> {

    }
  `
  },
  {
    name: 'ts different React components',
    typeSystem: 'typescript',
    code: `
    import {Component} from 'react';

    class FooComponent extends Component<{foo: void}> {

    }

    class BarComponent extends React.Component<{foo: void}> {

    }
  `
  },
  {
    name: 'ts tuple',
    typeSystem: 'typescript',
    code: `
    class Component extends React.Component<{foo: [string, number]}> {

    }
  `
  },
  {
    name: 'ts enum',
    typeSystem: 'typescript',
    code: `
    enum Color {Red, Green, Blue};
    class Component extends React.Component<{foo: Color}> {

    }
  `
  },

  {
    name: 'ts type alias',
    typeSystem: 'typescript',
    code: `
      type Props = {
        a: string;
      }

      class Component extends React.Component<Props> {
      }
    `
  },
  {
    name: 'ts type alias declaration - 1',
    typeSystem: 'typescript',
    code: `
      type LiteralType = 'one' | 'two';
      interface Props {
        a: LiteralType;
      }

      class Component extends React.Component<Props> {
      }
    `
  },
  {
    name: 'ts type alias declaration - 2',
    typeSystem: 'typescript',
    code: `
      type ReactElement = React.ReactElement<any> | React.ReactElement<any>[];
      interface Props {
        a: ReactElement;
      }

      class Component extends React.Component<Props> {
      }
    `
  },
  {
    name: 'Should handle importing other ts files',
    typeSystem: 'typescript',
    code: `
      import { Props } from "./__fixtures__/props";

      class Component extends React.Component<Props> {
      }
    `
  },
  {
    name: 'ts method signature',
    typeSystem: 'typescript',
    code: `
      interface Props {
        on(x: string): void;
      }

      class Component extends React.Component<Props> {}
    `
  },
  {
    name: 'ts call signature declaration',
    typeSystem: 'typescript',
    code: `
      interface Props {
        (x: string): void;
      }

      class Component extends React.Component<Props> {}
    `
  },
  {
    name: 'ts parenthesized type',
    typeSystem: 'typescript',
    code: `
      interface Props {
        a: ({ b: string })
      }

      class Component extends React.Component<Props> {}
    `
  },
  {
    name: 'ts interface extend',
    typeSystem: 'typescript',
    code: `
      import { Type } from './__fixtures__/types';
      interface DefaultTypes {
        b: number
      }

      interface Props extends DefaultTypes, Type {
        a: string
      }

      class Component extends React.Component<Props> {
        static defaultProps = {
          b: 1
        }
      }
    `
  },
  {
    name: 'ts export all',
    typeSystem: 'typescript',
    code: `
      import { NestedInterface1 } from './__fixtures__/types';

      class Component extends React.Component<NestedInterface1> {}
    `
  },
  {
    name: 'ts recursive type',
    typeSystem: 'typescript',
    code: `
      interface SiblingInterface {
        property: string
      }

      interface RecursiveType {
        properties: SiblingInterface
        type: RecursiveType
      }

      class Component extends React.Component<RecursiveType> {}
    `
  },

  {
    name: 'indexed type',
    typeSystem: 'typescript',
    code: `
      type MyType = {
        props: number;
      }

      class Component extends React.Component<{ foo: MyType['props'] }> {}
    `
  },
  {
    name: 'indexed imported type',
    typeSystem: 'typescript',
    code: `
      class Component extends React.Component<{ foo: ImportedType['props'] }> {}
    `
  },
  {
    name: 'generic + type expression',
    typeSystem: 'typescript',
    code: `
    type Foo<T, U> = T extends object ? T : U;

    class Component extends React.Component<Foo<string, {}>> {
    }
  `
  },
  {
    name: 'unknown prop',
    typeSystem: 'typescript',
    code: `
    import React from 'react';

    type Props = {
      children: unknown,
    };

    class Component extends React.Component<Props> {
    }
  `
  },
  {
    name: 'follow export default export',
    typeSystem: 'typescript',
    code: `
      export { default } from './__fixtures__/component.tsx';
    `
  },
  {
    name: 'follow export named default export',
    typeSystem: 'typescript',
    code: `
      export { x as default } from './__fixtures__/componentB.tsx';
    `
  },
  {
    name: 'type query for class component',
    typeSystem: 'typescript',
    code: `
    import React from 'react';

    import { NestedInterface1 } from './__fixtures__/types';
    import { NestedInterface2 } from 'external-package';

    interface Props extends NestedInterface1, NestedInterface2 {
      foo: number;
      bar?: string;
      baz?: boolean
    }

    const value = {
      a: 'a value',
      b: 'b value',
    };

    type MyComponentProps = Props & typeof MyComponent.defaultProps & typeof value;

    class MyComponent extends React.Component<MyComponentProps> {
      static defaultProps = {
        bar: 'bar',
        baz: true
      }
    }
  `
  },
  {
    name: 'type query for functional component',
    typeSystem: 'typescript',
    code: `
    import React from 'react';

    type OCProps = {
      type: 'a' | 'b' | 'c' | 'd',
    }
    function OtherComponent(props: OCProps) {}

    type MCProps = {
      type: React.ComponentProps<typeof OtherComponent>['type'],
    }
    function MyComponent(props: MCProps) {}

    export default MyComponent;
  `
  },
  {
    name: 'forwardRef typed via generic types',
    typeSystem: 'typescript',
    code: `
    import React, { forwardRef } from 'react';

    type MyComponentProps = {
      foo: string,
    }

    const MyComponent = forwardRef<HTMLElement, MyComponentProps>((props, ref) => {
      return <span>Foo</span>;
    });

    export default MyComponent;
  `
  },
  {
    name: 'forwardRef typed via function args',
    typeSystem: 'typescript',
    code: `
    import React, { forwardRef } from 'react';

    type MyComponentProps = {
      foo: string,
    }

    const MyComponent = forwardRef((props: MyComponentProps, ref: HTMLElement) => {
      return <span>Foo</span>;
    });

    export default MyComponent;
  `
  }
];

cases(
  'TypeScript:',
  testCase => {
    let code = stripIndent(testCase.code);
    // Pass in file name so we can resolve imports to files in __fixtures__
    let result = extractReactTypes(code, testCase.typeSystem, __filename);
    expect(result).toMatchSnapshot();
  },
  TESTS
);
