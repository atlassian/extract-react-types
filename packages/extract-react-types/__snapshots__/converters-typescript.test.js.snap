// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should handle importing other ts files 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "id",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "number",
          },
        },
        Object {
          "key": Object {
            "kind": "id",
            "name": "name",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "string",
          },
        },
      ],
    },
  },
  "kind": "program",
}
`;

exports[`follow export default export 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "className",
          },
          "kind": "property",
          "optional": true,
          "value": Object {
            "kind": "string",
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`follow export named default export 1`] = `
Object {
  "component": undefined,
  "kind": "program",
}
`;

exports[`ts any 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "any",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts array 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "key": Object {
            "kind": "any",
          },
          "kind": "generic",
          "typeParams": Object {
            "kind": "typeParams",
            "params": Array [
              Object {
                "kind": "number",
              },
            ],
          },
          "value": Object {
            "kind": "id",
            "name": "Array",
          },
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts array prop 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Schedule",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "intervals",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "key": Object {
              "kind": "any",
            },
            "kind": "generic",
            "typeParams": Object {
              "kind": "typeParams",
              "params": Array [
                Object {
                  "kind": "object",
                  "members": Array [
                    Object {
                      "key": Object {
                        "kind": "id",
                        "name": "begin",
                      },
                      "kind": "property",
                      "optional": false,
                      "value": Object {
                        "kind": "generic",
                        "value": Object {
                          "kind": "id",
                          "name": "Interval['begin']",
                        },
                      },
                    },
                    Object {
                      "key": Object {
                        "kind": "id",
                        "name": "end",
                      },
                      "kind": "property",
                      "optional": false,
                      "value": Object {
                        "kind": "generic",
                        "value": Object {
                          "kind": "id",
                          "name": "Interval['end']",
                        },
                      },
                    },
                  ],
                },
              ],
            },
            "value": Object {
              "kind": "id",
              "name": "Array",
            },
          },
        },
      ],
      "referenceIdName": "ScheduleProps",
    },
  },
  "kind": "program",
}
`;

exports[`ts as expression 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "default": Object {
            "kind": "string",
            "value": "foo",
          },
          "key": Object {
            "kind": "id",
            "name": "bar",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "generic",
            "value": Object {
              "kind": "union",
              "referenceIdName": "Foo",
              "types": Array [
                Object {
                  "kind": "string",
                  "value": "foo",
                },
                Object {
                  "kind": "string",
                  "value": "bar",
                },
              ],
            },
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts boolean 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "boolean",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts call signature declaration 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "string",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "param",
                "type": null,
                "value": Object {
                  "kind": "id",
                  "name": "x",
                },
              },
            ],
            "returnType": Object {
              "kind": "void",
            },
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts custom prop 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Badge",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "texture",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "string",
          },
        },
      ],
      "referenceIdName": "BadgeProps",
    },
  },
  "kind": "program",
}
`;

exports[`ts custom prop 2`] = `
Object {
  "component": Object {
    "key": Object {
      "kind": "any",
    },
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "typeParams": Object {
      "kind": "typeParams",
      "params": Array [
        Object {
          "kind": "string",
        },
        Object {
          "kind": "object",
          "members": Array [],
        },
      ],
    },
    "value": Object {
      "kind": "any",
      "referenceIdName": "Foo",
    },
  },
  "kind": "program",
}
`;

exports[`ts decorators 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [],
    "name": Object {
      "kind": "id",
      "name": "Theme",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts different React components 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "void",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "FooComponent",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts enum 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "generic",
          "value": Object {
            "kind": "union",
            "referenceIdName": "Color",
            "types": Array [
              Object {
                "kind": "id",
                "name": "Color.Red",
              },
              Object {
                "kind": "id",
                "name": "Color.Green",
              },
              Object {
                "kind": "id",
                "name": "Color.Blue",
              },
            ],
          },
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts export all 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "nestedProperty1",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "boolean",
          },
        },
        Object {
          "key": Object {
            "kind": "id",
            "name": "nestedProperty2",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "string",
          },
        },
      ],
    },
  },
  "kind": "program",
}
`;

exports[`ts function 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "generic",
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "param",
                "type": null,
                "value": Object {
                  "kind": "id",
                  "name": "string",
                },
              },
              Object {
                "kind": "param",
                "type": null,
                "value": Object {
                  "kind": "id",
                  "name": "boolean",
                },
              },
            ],
            "returnType": Object {
              "kind": "number",
            },
          },
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts interface 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "foo",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "string",
          },
        },
        Object {
          "key": Object {
            "kind": "id",
            "name": "bar",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "number",
          },
        },
      ],
      "referenceIdName": "ComponentProps",
    },
  },
  "kind": "program",
}
`;

exports[`ts interface extend 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "a",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "string",
          },
        },
        Object {
          "default": Object {
            "kind": "number",
            "value": 1,
          },
          "key": Object {
            "kind": "id",
            "name": "b",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "number",
          },
        },
        Object {
          "key": Object {
            "kind": "id",
            "name": "c",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "boolean",
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts literals 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "union",
          "types": Array [
            Object {
              "kind": "string",
              "value": "25",
            },
            Object {
              "kind": "string",
              "value": 30,
            },
          ],
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts method signature 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "on",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "function",
            "parameters": Array [
              Object {
                "kind": "param",
                "type": null,
                "value": Object {
                  "kind": "id",
                  "name": "x",
                },
              },
            ],
            "returnType": Object {
              "kind": "void",
            },
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts nested prop 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Badge",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "texture",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "generic",
            "value": Object {
              "kind": "id",
              "name": "Texture['src']",
            },
          },
        },
      ],
      "referenceIdName": "BadgeProps",
    },
  },
  "kind": "program",
}
`;

exports[`ts no react component 1`] = `
Object {
  "component": undefined,
  "kind": "program",
}
`;

exports[`ts object 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "object",
          "members": Array [
            Object {
              "key": Object {
                "kind": "id",
                "name": "name",
              },
              "kind": "property",
              "optional": false,
              "value": Object {
                "kind": "string",
              },
            },
          ],
        },
      },
      Object {
        "key": Object {
          "kind": "id",
          "name": "bar",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "number",
        },
      },
      Object {
        "key": Object {
          "kind": "id",
          "name": "verified",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "boolean",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts optional 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": true,
        "value": Object {
          "kind": "string",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts parenthesized type 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "a",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "object",
            "members": Array [
              Object {
                "key": Object {
                  "kind": "id",
                  "name": "b",
                },
                "kind": "property",
                "optional": false,
                "value": Object {
                  "kind": "string",
                },
              },
            ],
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts recursive type 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "properties",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "generic",
            "value": Object {
              "kind": "object",
              "members": Array [
                Object {
                  "key": Object {
                    "kind": "id",
                    "name": "property",
                  },
                  "kind": "property",
                  "optional": false,
                  "value": Object {
                    "kind": "string",
                  },
                },
              ],
              "referenceIdName": "SiblingInterface",
            },
          },
        },
        Object {
          "key": Object {
            "kind": "id",
            "name": "type",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "generic",
            "value": Object {
              "kind": "id",
              "name": "RecursiveType",
            },
          },
        },
      ],
      "referenceIdName": "RecursiveType",
    },
  },
  "kind": "program",
}
`;

exports[`ts string 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "string",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts tuple 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "tuple",
          "types": Array [
            Object {
              "kind": "string",
            },
            Object {
              "kind": "number",
            },
          ],
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts type alias 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "a",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "string",
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts type alias declaration - 1 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "a",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "generic",
            "value": Object {
              "kind": "union",
              "referenceIdName": "LiteralType",
              "types": Array [
                Object {
                  "kind": "string",
                  "value": "one",
                },
                Object {
                  "kind": "string",
                  "value": "two",
                },
              ],
            },
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts type alias declaration - 2 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "a",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "generic",
            "value": Object {
              "kind": "union",
              "referenceIdName": "ReactElement",
              "types": Array [
                Object {
                  "key": Object {
                    "kind": "any",
                  },
                  "kind": "generic",
                  "typeParams": Object {
                    "kind": "typeParams",
                    "params": Array [
                      Object {
                        "kind": "any",
                      },
                    ],
                  },
                  "value": Object {
                    "kind": "id",
                    "name": "React.ReactElement",
                  },
                },
                Object {
                  "kind": "arrayType",
                  "type": Object {
                    "key": Object {
                      "kind": "any",
                    },
                    "kind": "generic",
                    "typeParams": Object {
                      "kind": "typeParams",
                      "params": Array [
                        Object {
                          "kind": "any",
                        },
                      ],
                    },
                    "value": Object {
                      "kind": "id",
                      "name": "React.ReactElement",
                    },
                  },
                },
              ],
            },
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts union 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "union",
          "types": Array [
            Object {
              "kind": "number",
            },
            Object {
              "kind": "string",
            },
          ],
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`ts unknown prop 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "children",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "unknown",
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`ts void 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "void",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`typescript React.ComponentType 1`] = `
Object {
  "component": Object {
    "kind": "generic",
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
    "value": Object {
      "kind": "object",
      "members": Array [
        Object {
          "key": Object {
            "kind": "id",
            "name": "children",
          },
          "kind": "property",
          "optional": false,
          "value": Object {
            "kind": "generic",
            "value": Object {
              "kind": "id",
              "name": "React.ComponentType",
            },
          },
        },
      ],
      "referenceIdName": "Props",
    },
  },
  "kind": "program",
}
`;

exports[`typescript indexed imported type 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "generic",
          "value": Object {
            "kind": "id",
            "name": "ImportedType['props']",
          },
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;

exports[`typescript indexed type 1`] = `
Object {
  "component": Object {
    "kind": "object",
    "members": Array [
      Object {
        "key": Object {
          "kind": "id",
          "name": "foo",
        },
        "kind": "property",
        "optional": false,
        "value": Object {
          "kind": "number",
        },
      },
    ],
    "name": Object {
      "kind": "id",
      "name": "Component",
      "type": null,
    },
  },
  "kind": "program",
}
`;
